src/main/java/com/google/finapp/SpannerDaoJDBCImpl.java

```diff
@@ -0,0 +1,176 @@
+    }
+  }
+
+  public void createAccount(
+      ByteArray accountId, AccountType accountType, AccountStatus accountStatus, BigDecimal balance)
+      throws SpannerDaoException {
+    try (Connection connection = DriverManager.getConnection(this.connectionUrl);
+        PreparedStatement ps =
+            connection.prepareStatement(
+                "INSERT INTO Account\n"
+                    + "(AccountId, AccountType, AccountStatus, Balance, CreationTimestamp)\n"
+                    + "VALUES\n"
+                    + "(?, ?, ?, ?, PENDING_COMMIT_TIMESTAMP())")) {
+      ps.setBytes(1, accountId.toByteArray());
+      ps.setInt(2, accountType.getNumber());
+      ps.setInt(3, accountStatus.getNumber());
+      ps.setBigDecimal(4, balance);
+      ps.executeUpdate();
+    } catch (SQLException e) {
+      throw new SpannerDaoException(e);
+    }
+  }
+
+  public void addAccountForCustomer(
+      ByteArray customerId, ByteArray accountId, ByteArray roleId, String roleName)
+      throws SpannerDaoException {
+    try (Connection connection = DriverManager.getConnection(this.connectionUrl);
+        PreparedStatement ps =
+            connection.prepareStatement(
+                "INSERT INTO CustomerRole\n"
+                    + "(CustomerId, AccountId, RoleId, Role)\n"
+                    + "VALUES\n"
+                    + "(?, ?, ?, ?)")) {
+      ps.setBytes(1, customerId.toByteArray());
+      ps.setBytes(2, accountId.toByteArray());
+      ps.setBytes(3, roleId.toByteArray());
+      ps.setString(4, roleName);
+      ps.executeUpdate();
+    } catch (SQLException e) {
+      throw new SpannerDaoException(e);
+    }
+  }
+
+  public void moveAccountBalance(ByteArray fromAccountId, ByteArray toAccountId, BigDecimal amount)
+      throws SpannerDaoException {
+    try (Connection connection = DriverManager.getConnection(this.connectionUrl);
+        PreparedStatement readStatement =
+            connection.prepareStatement(
+                "SELECT AccountId, Balance FROM Account WHERE (AccountId = ? or AccountId = ?)")) {
+      connection.setAutoCommit(false);
+      byte[] fromAccountIdArray = fromAccountId.toByteArray();
+      byte[] toAccountIdArray = toAccountId.toByteArray();
+      readStatement.setBytes(1, fromAccountIdArray);
+      readStatement.setBytes(2, toAccountIdArray);
+      java.sql.ResultSet resultSet = readStatement.executeQuery();
+      BigDecimal sourceAmount = new BigDecimal(Long.MIN_VALUE);
+      BigDecimal destAmount = new BigDecimal(Long.MIN_VALUE);
+      while (resultSet.next()) {
+        byte[] currentId = resultSet.getBytes("AccountId");
+        if (Arrays.equals(currentId, fromAccountIdArray)) {
+          sourceAmount = resultSet.getBigDecimal("Balance");
+        } else {
+          destAmount = resultSet.getBigDecimal("Balance");
+        }
+      }
+      if (sourceAmount.min(destAmount).equals(new BigDecimal(Long.MIN_VALUE))) {
+        throw new IllegalArgumentException();
```

Comment:
Since I initialized the values to be `Big Decimal(Long.MIN_VALUE)`, I figured I could utilize that here to see if either the fromAccount or toAccount weren't found in the database. We haven't formally established this, but we could have an invariant where there are no negative balances in the database. With that invariant, we could set the `BigDecimal` value to -1 or something similar to simplify things. Thoughts @voulg ?

Reply:
fyi `(new BigDecimal(Long.MIN_VALUE)).equals(new BigDecimal(Long.MIN_VALUE))` --> `true`